src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -589,8 +594,23 @@ InterpreterSelectQuery::InterpreterSelectQuery(
 
 void InterpreterSelectQuery::buildQueryPlan(QueryPlan & query_plan)
 {
+    if (cached_data.contains(query_ptr->getTreeHash()))
+    {
+        auto &header= cached_data[query_ptr->getTreeHash()].first;
+        const auto &chunks = cached_data[query_ptr->getTreeHash()].second;
+
+        Pipe pipe(std::make_shared<ReadFromCacheTransform>(header, chunks));
+        auto read_from_cache_step = std::make_unique<ReadFromPreparedSource>(std::move(pipe));
+        read_from_cache_step->setStepDescription("Read query result from cache");
+        query_plan.addStep(std::move(read_from_cache_step));
+        return;
+    }
     executeImpl(query_plan, std::move(input_pipe));
 
+    auto caching_step = std::make_unique<CachingStep>(query_plan.getCurrentDataStream(), cached_data, query_ptr);
```

Comment:
Currently better to only use cache if some setting is specified, for now we can make simple `use_query_cache`, check `Settings.h`.

Reply:
Added two options for that: query_cache_active_usage and query_cache_passive_usage.